File: src/main/java/org/janelia/saalfeldlab/paintera/config/BookmarkConfigNode.java
package org.janelia.saalfeldlab.paintera.config;

import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.value.ChangeListener;
import javafx.collections.ListChangeListener;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Node;
import javafx.scene.control.Alert;
import javafx.scene.control.Button;
import javafx.scene.control.ButtonType;
import javafx.scene.control.ContentDisplay;
import javafx.scene.control.Label;
import javafx.scene.control.TextArea;
import javafx.scene.control.TextField;
import javafx.scene.control.TitledPane;
import javafx.scene.control.Tooltip;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import javafx.scene.transform.Affine;
import javafx.util.Duration;
import javafx.util.Pair;
import net.imglib2.realtransform.AffineTransform3D;
import org.janelia.saalfeldlab.fx.Labels;
import org.janelia.saalfeldlab.fx.TitledPanes;
import org.janelia.saalfeldlab.fx.ui.NumericSliderWithField;
import org.janelia.saalfeldlab.paintera.ui.PainteraAlerts;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.invoke.MethodHandles;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.stream.IntStream;

public class BookmarkConfigNode extends TitledPane {

	private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

	private static class BookmarkNode extends VBox {

		private final TextArea label;

		private BookmarkNode(final BookmarkConfig.Bookmark bookmark) {
			label = new TextArea(bookmark.getNote());
			label.setEditable(false);
			label.setWrapText(false);
			label.setTooltip(new Tooltip(label.getText()));

			final Node globalTransformGrid = affineTransformGrid(bookmark.getGlobalTransformCopy());
			final TitledPane globalTransformPane = TitledPanes.createCollapsed("Global Transform", globalTransformGrid);
			globalTransformPane.setPadding(Insets.EMPTY);


			final Node viewer3DTransformGrid = viewer3DTransformGrid(bookmark.getViewer3DTransformCopy());
			final TitledPane viewer3DTransformPane = TitledPanes.createCollapsed("3D Viewer Transform", viewer3DTransformGrid);
			viewer3DTransformPane.setPadding(Insets.EMPTY);

			getChildren().addAll(label, globalTransformPane, viewer3DTransformPane);
		}

	}

	private static class BookmarkTitledPane extends TitledPane {
		private BookmarkTitledPane(
				final BookmarkConfig.Bookmark bookmark,
				final int id,
				final BiConsumer<BookmarkConfig.Bookmark, BookmarkConfig.Bookmark> replace,
				final Consumer<BookmarkConfig.Bookmark> onApply,
				final Consumer<BookmarkConfig.Bookmark> onRemove) {
			super(null, null);
			final int oneBasedId = id + 1;
			setText("" + oneBasedId);
			final Button goThere = new Button("Apply");
			final Button closeIt = new Button("x");
			final Button updateNote = new Button("Update Note");

			goThere.setOnAction(e -> onApply.accept(bookmark));
			closeIt.setOnAction(e -> {
				final Alert dialog = bookmarkDialog(bookmark).getKey();
				dialog.setHeaderText("Remove bookmark " + oneBasedId + "?");
				if (ButtonType.OK.equals(dialog.showAndWait().orElse(ButtonType.CANCEL)))
					onRemove.accept(bookmark);
			});
			updateNote.setOnAction(e -> {
				final Pair<Alert, TextArea> dialog = bookmarkDialog(bookmark);
				dialog.getKey().setHeaderText("Update Bookmark Note");
				if (ButtonType.OK.equals(dialog.getKey().showAndWait().orElse(ButtonType.CANCEL)))
					replace.accept(bookmark, bookmark.withNote(dialog.getValue().getText()));
			});

			closeIt.setTooltip(new Tooltip("Remove bookmark " + oneBasedId));

			final Label noteLabel = Labels.withTooltip(Optional.ofNullable(bookmark.getNote()).map(n -> n.replace("\n", " ")).orElse(null));
			noteLabel.setPrefWidth(100.0);
			final HBox hBox = new HBox(noteLabel, goThere, updateNote, closeIt);
			hBox.setAlignment(Pos.CENTER);

			setGraphic(hBox);
			setPadding(Insets.EMPTY);
			setExpanded(false);
			setContent(new BookmarkNode(bookmark));
			setContentDisplay(ContentDisplay.RIGHT);
		}

	}

	private static TextField matrixField(final String text) {
		final TextField tf = new TextField(text);
		tf.setMinWidth(30.0);
		tf.setEditable(false);
		return tf;
	}

	private static Node affineTransformGrid(final AffineTransform3D transform) {
		final GridPane transformGrid = new GridPane();
		for (int i = 0; i < 3; ++i)
			for (int k = 0; k < 4; ++k)
				transformGrid.add(matrixField(Double.toString(transform.get(i, k))), k, i);
		transformGrid.setPadding(Insets.EMPTY);
		return transformGrid;
	}

	private static Node viewer3DTransformGrid(final Affine transform) {
		final GridPane transformGrid = new GridPane();
		transformGrid.add(matrixField(Double.toString(transform.getMxx())),0, 0);
		transformGrid.add(matrixField(Double.toString(transform.getMxy())),1, 0);
		transformGrid.add(matrixField(Double.toString(transform.getMxz())),2, 0);
		transformGrid.add(matrixField(Double.toString(transform.getMyx())),0, 1);
		transformGrid.add(matrixField(Double.toString(transform.getMyy())),1, 1);
		transformGrid.add(matrixField(Double.toString(transform.getMyz())),2, 1);
		transformGrid.add(matrixField(Double.toString(transform.getMzx())),0, 2);
		transformGrid.add(matrixField(Double.toString(transform.getMzy())),1, 2);
		transformGrid.add(matrixField(Double.toString(transform.getMzz())),2, 2);

		transformGrid.add(matrixField(Double.toString(transform.getTx())),3, 0);
		transformGrid.add(matrixField(Double.toString(transform.getTx())),3, 1);
		transformGrid.add(matrixField(Double.toString(transform.getTx())),3, 2);

		transformGrid.setPadding(Insets.EMPTY);
		return transformGrid;
	}

	private final Consumer<BookmarkConfig.Bookmark> applyBookmark;

	private final ObjectProperty<BookmarkConfig> bookmarkConfig = new SimpleObjectProperty<>();

	private final ObjectProperty<Duration> transitionTime = new SimpleObjectProperty<>(Duration.millis(300.0));

	private final NumericSliderWithField transitionTimeSlider = new NumericSliderWithField(0.0, 1000.0, 300.0);

	private final VBox bookmarkSettings = new VBox(new HBox(new Label("Transition Time"), transitionTimeSlider.slider(), transitionTimeSlider.textField()));

	private final VBox bookmarkNodes = new VBox();

	private final ChangeListener<BookmarkConfig> configListener = (obs, oldv, newv) -> {
		if (oldv != null) {
			oldv.getUnmodifiableBookmarks().removeListener(this.listListener);
			this.transitionTime.unbindBidirectional(oldv.transitionTimeProperty());
		}
		if (newv !=null) {
			newv.getUnmodifiableBookmarks().addListener(this.listListener);
			this.transitionTime.bindBidirectional(newv.transitionTimeProperty());
			this.transitionTime.set(newv.transitionTimeProperty().get());
			updateChildren(new ArrayList<>(newv.getUnmodifiableBookmarks()), newv::replaceBookmark, newv::removeBookmark);
		}
	};

	private final ListChangeListener<BookmarkConfig.Bookmark> listListener = change -> updateChildren(
			change.getList(),
			bookmarkConfig.get()::replaceBookmark,
			bookmarkConfig.get()::removeBookmark);

	public BookmarkConfigNode(final Consumer<BookmarkConfig.Bookmark> applyBookmark) {
		super("Bookmarks", null);
		this.applyBookmark = applyBookmark;

		this.transitionTimeSlider.slider().valueProperty().addListener((obs, oldv, newv) -> this.transitionTime.set(Duration.millis(newv.doubleValue())));
		this.transitionTime.addListener((obs, oldv, newv) -> this.transitionTimeSlider.slider().setValue(newv.toMillis()));

		setExpanded(false);
		this.bookmarkConfig.addListener(configListener);
		bookmarkSettings.setPadding(Insets.EMPTY);
		bookmarkNodes.setPadding(Insets.EMPTY);
		setContent(new VBox(bookmarkSettings, bookmarkNodes));
	}

	private void updateChildren(
			final List<? extends BookmarkConfig.Bookmark> bookmarks,
			final BiConsumer<BookmarkConfig.Bookmark, BookmarkConfig.Bookmark> replaceBookmark,
			final Consumer<BookmarkConfig.Bookmark> removeBookmark) {
		LOG.debug("Updating contents with {}", bookmarks);
		final Node[] nodes = IntStream
				.range(0, bookmarks.size())
				.mapToObj(i -> new BookmarkTitledPane(bookmarks.get(i), i, replaceBookmark, applyBookmark, removeBookmark))
				.toArray(Node[]::new);
		bookmarkNodes.getChildren().setAll(nodes);
	}

	public ObjectProperty<BookmarkConfig> bookmarkConfigProperty() {
		return this.bookmarkConfig;
	}

	public void setBookmarkConfig(final BookmarkConfig bookmarkConfig) {
		bookmarkConfigProperty().set(bookmarkConfig);
	}

	public BookmarkConfig getBookmarkConfig() {
		return bookmarkConfigProperty().get();
	}

	public void requestAddNewBookmark(
			final AffineTransform3D globalTransform,
			final Affine viewer3DTransform) {

		final Pair<Alert, TextArea> dialog = bookmarkDialog(globalTransform, viewer3DTransform, null);
		dialog.getKey().setHeaderText("Bookmark current view");

		final Optional<ButtonType> bt = dialog.getKey().showAndWait();

		if (ButtonType.OK.equals(bt.orElse(ButtonType.CANCEL))) {
			bookmarkConfig.get().addBookmark(new BookmarkConfig.Bookmark(
					globalTransform,
					viewer3DTransform,
					dialog.getValue().getText()));
		}

	}

	private static Pair<Alert, TextArea> bookmarkDialog(
			final BookmarkConfig.Bookmark bookmark) {
		return bookmarkDialog(bookmark.getGlobalTransformCopy(), bookmark.getViewer3DTransformCopy(), bookmark.getNote());
	}

	private static Pair<Alert, TextArea> bookmarkDialog(
			final AffineTransform3D globalTransform,
			final Affine viewer3DTransform,
			final String note) {
		final Alert dialog = PainteraAlerts.alert(Alert.AlertType.CONFIRMATION, true);
		final TextArea label = new TextArea(note);

		dialog.getDialogPane().setContent(new VBox(
				label,
				TitledPanes.createCollapsed("Global Transform", affineTransformGrid(globalTransform)),
				TitledPanes.createCollapsed("3D Viewer Transform", viewer3DTransformGrid(viewer3DTransform))));
		return new Pair<>(dialog, label);

	}

}

File: src/main/kotlin/org/janelia/saalfeldlab/paintera/config/BookmarkConfigNode.kt
package org.janelia.saalfeldlab.paintera.config

import javafx.beans.property.SimpleObjectProperty
import javafx.beans.value.ChangeListener
import javafx.collections.ListChangeListener
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.scene.Node
import javafx.scene.control.*
import javafx.scene.layout.GridPane
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import javafx.scene.transform.Affine
import javafx.util.Duration
import javafx.util.Pair
import net.imglib2.realtransform.AffineTransform3D
import org.janelia.saalfeldlab.fx.Labels
import org.janelia.saalfeldlab.fx.TitledPanes
import org.janelia.saalfeldlab.fx.ui.MarkdownPane
import org.janelia.saalfeldlab.fx.ui.NumericSliderWithField
import org.janelia.saalfeldlab.fx.util.InvokeOnJavaFXApplicationThread
import org.janelia.saalfeldlab.paintera.Paintera
import org.janelia.saalfeldlab.paintera.ui.dialogs.PainteraAlerts
import org.slf4j.LoggerFactory
import java.lang.invoke.MethodHandles
import java.util.Optional
import java.util.function.BiConsumer
import java.util.function.Consumer

class BookmarkConfigNode private constructor(private val applyBookmark: (BookmarkConfig.Bookmark) -> Unit) : TitledPane("Bookmarks", null) {

	constructor(bookmarkConfig: BookmarkConfig, applyBookmark: (BookmarkConfig.Bookmark) -> Unit) : this(applyBookmark) {
		Paintera.whenPaintable {
			InvokeOnJavaFXApplicationThread {
				this@BookmarkConfigNode.bookmarkConfig.set(bookmarkConfig)
			}
		}
	}

	// TODO change this into a regular bookmark
	private val bookmarkConfig = SimpleObjectProperty<BookmarkConfig>()

	private val transitionTime = SimpleObjectProperty(Duration.millis(300.0))

	private val transitionTimeSlider = NumericSliderWithField(0.0, 1000.0, 300.0)

	private val bookmarkSettings = VBox(HBox(Label("Transition Time"), transitionTimeSlider.slider, transitionTimeSlider.textField))

	private val bookmarkNodes = VBox()

	private val configListener = ChangeListener<BookmarkConfig> { _, oldv, newv ->
		oldv?.let {
			it.unmodifiableBookmarks.removeListener(this.listListener)
			this.transitionTime.unbindBidirectional(it.transitionTimeProperty())
		}
		newv?.let {
			it.unmodifiableBookmarks.addListener(this.listListener)
			this.transitionTime.bindBidirectional(it.transitionTimeProperty())
			this.transitionTime.set(it.transitionTimeProperty().get())
			updateChildren(
				it.unmodifiableBookmarks,
				{ replaced, with -> it.replaceBookmark(replaced, with) },
				{ bm -> it.removeBookmark(bm) })
		}
	}

	private val listListener = ListChangeListener<BookmarkConfig.Bookmark> {
		updateChildren(
			it.getList(),
			{ replaced, with -> bookmarkConfig.get().replaceBookmark(replaced, with) },
			{ bm -> bookmarkConfig.get().removeBookmark(bm) })
	}

	private class BookmarkNode constructor(bookmark: BookmarkConfig.Bookmark) : VBox() {

		private val markdownNode = MarkdownPane().also { it.text = bookmark.note }

		init {
			markdownNode.isEditable = false
			markdownNode.isWrapText = false
			markdownNode.tooltip = Tooltip(markdownNode.text)
			markdownNode.showRenderedTab()

			val globalTransformGrid = affineTransformGrid(bookmark.globalTransformCopy)
			val globalTransformPane = TitledPanes.createCollapsed("Global Transform", globalTransformGrid)
			globalTransformPane.padding = Insets.EMPTY


			val viewer3DTransformGrid = viewer3DTransformGrid(bookmark.viewer3DTransformCopy)
			val viewer3DTransformPane = TitledPanes.createCollapsed("3D Viewer Transform", viewer3DTransformGrid)
			viewer3DTransformPane.padding = Insets.EMPTY

			children.addAll(markdownNode, globalTransformPane, viewer3DTransformPane)
		}

	}

	private class BookmarkTitledPane constructor(
		bookmark: BookmarkConfig.Bookmark,
		id: Int,
		replace: BiConsumer<BookmarkConfig.Bookmark, BookmarkConfig.Bookmark>,
		onApply: Consumer<BookmarkConfig.Bookmark>,
		onRemove: Consumer<BookmarkConfig.Bookmark>
	) : TitledPane(null, null) {
		init {
			val oneBasedId = id + 1
			text = "" + oneBasedId
			val goThere = Button("Apply")
			val closeIt = Button("x")
			val updateNote = Button("Update Note")

			goThere.setOnAction { onApply.accept(bookmark) }
			closeIt.setOnAction {
				val dialogAndMarkdown = bookmarkDialog(bookmark)
				val dialog = dialogAndMarkdown.key
				dialogAndMarkdown.value.isEditable = false
				dialogAndMarkdown.value.showRenderedTab()
				dialog.headerText = "Remove bookmark $oneBasedId?"
				if (ButtonType.OK == dialog.showAndWait().orElse(ButtonType.CANCEL))
					onRemove.accept(bookmark)
			}
			updateNote.setOnAction {
				val dialog = bookmarkDialog(bookmark)
				dialog.key.headerText = "Update Bookmark Note"
				if (ButtonType.OK == dialog.key.showAndWait().orElse(ButtonType.CANCEL))
					replace.accept(bookmark, bookmark.withNote(dialog.value.text ?: ""))
			}

			closeIt.tooltip = Tooltip("Remove bookmark $oneBasedId")

			val noteLabel = Labels.withTooltip(Optional.ofNullable(bookmark.note).map { n -> n.replace("\n", " ") }.orElse(""))
			noteLabel.prefWidth = 100.0
			val hBox = HBox(noteLabel, goThere, updateNote, closeIt)
			hBox.alignment = Pos.CENTER

			graphic = hBox
			padding = Insets.EMPTY
			isExpanded = false
			content = BookmarkNode(bookmark)
			contentDisplay = ContentDisplay.RIGHT
		}

	}

	init {

		this.transitionTimeSlider.slider.valueProperty().addListener { _, _, newv -> this.transitionTime.set(Duration.millis(newv.toDouble())) }
		this.transitionTime.addListener { _, _, newv -> this.transitionTimeSlider.slider.value = newv.toMillis() }

		isExpanded = false
		this.bookmarkConfig.addListener(configListener)
		bookmarkSettings.padding = Insets.EMPTY
		bookmarkNodes.padding = Insets.EMPTY
		content = VBox(bookmarkSettings, bookmarkNodes)
	}

	private fun updateChildren(
		bookmarks: List<BookmarkConfig.Bookmark>,
		replaceBookmark: BiConsumer<BookmarkConfig.Bookmark, BookmarkConfig.Bookmark>,
		removeBookmark: Consumer<BookmarkConfig.Bookmark>
	) {
		LOG.debug("Updating contents with {}", bookmarks)
		val nodes = Array(bookmarks.size) { BookmarkTitledPane(bookmarks[it], it, replaceBookmark, applyBookmark, removeBookmark) }
		bookmarkNodes.children.setAll(*nodes)
	}

	fun requestAddNewBookmark(
		globalTransform: AffineTransform3D,
		viewer3DTransform: Affine
	) {

		val dialog = bookmarkDialog(globalTransform, viewer3DTransform, null)
		dialog.key.headerText = "Bookmark current view"

		val bt = dialog.key.showAndWait()

		if (ButtonType.OK == bt.orElse(ButtonType.CANCEL)) {
			bookmarkConfig.get().addBookmark(
				BookmarkConfig.Bookmark(
					globalTransform,
					viewer3DTransform,
					dialog.value.text
				)
			)
		}

	}

	companion object {

		private val LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass())

		private fun matrixField(text: String): TextField {
			val tf = TextField(text)
			tf.minWidth = 30.0
			tf.isEditable = false
			return tf
		}

		private fun affineTransformGrid(transform: AffineTransform3D): Node {
			val transformGrid = GridPane()
			for (i in 0..2)
				for (k in 0..3)
					transformGrid.add(matrixField(java.lang.Double.toString(transform.get(i, k))), k, i)
			transformGrid.padding = Insets.EMPTY
			return transformGrid
		}

		private fun viewer3DTransformGrid(transform: Affine): Node {
			val transformGrid = GridPane()
			transformGrid.add(matrixField(java.lang.Double.toString(transform.mxx)), 0, 0)
			transformGrid.add(matrixField(java.lang.Double.toString(transform.mxy)), 1, 0)
			transformGrid.add(matrixField(java.lang.Double.toString(transform.mxz)), 2, 0)
			transformGrid.add(matrixField(java.lang.Double.toString(transform.myx)), 0, 1)
			transformGrid.add(matrixField(java.lang.Double.toString(transform.myy)), 1, 1)
			transformGrid.add(matrixField(java.lang.Double.toString(transform.myz)), 2, 1)
			transformGrid.add(matrixField(java.lang.Double.toString(transform.mzx)), 0, 2)
			transformGrid.add(matrixField(java.lang.Double.toString(transform.mzy)), 1, 2)
			transformGrid.add(matrixField(java.lang.Double.toString(transform.mzz)), 2, 2)

			transformGrid.add(matrixField(java.lang.Double.toString(transform.tx)), 3, 0)
			transformGrid.add(matrixField(java.lang.Double.toString(transform.tx)), 3, 1)
			transformGrid.add(matrixField(java.lang.Double.toString(transform.tx)), 3, 2)

			transformGrid.padding = Insets.EMPTY
			return transformGrid
		}

		private fun bookmarkDialog(
			bookmark: BookmarkConfig.Bookmark
		): Pair<Alert, MarkdownPane> {
			return bookmarkDialog(bookmark.globalTransformCopy, bookmark.viewer3DTransformCopy, bookmark.note)
		}

		private fun bookmarkDialog(
			globalTransform: AffineTransform3D,
			viewer3DTransform: Affine,
			note: String?
		): Pair<Alert, MarkdownPane> {
			val dialog = PainteraAlerts.alert(Alert.AlertType.CONFIRMATION, true)
			val editor = MarkdownPane()
				.also { it.text = note }
				.also { it.showEditTab() }

			dialog.dialogPane.content = VBox(
				editor,
				TitledPanes.createCollapsed("Global Transform", affineTransformGrid(globalTransform)),
				TitledPanes.createCollapsed("3D Viewer Transform", viewer3DTransformGrid(viewer3DTransform))
			)
			return Pair(dialog, editor)

		}
	}

}
